### ISSUE TITLE
[feature] Build packages in parallel

### ISSUE THREAD

**Issue by blackliner** [conan-io/conan#9383]
Would it be possible for conan to build packages in parallel? Similar to `colcon`, which just shows the overall status (Like `building packages 4/18` ) conan should know the correct order from the overall build tree.

WHY? it would speed up the total process, since many small conan packages do not saturate higher core CPUs with building (cmake, packaging etc is all single core anyways)

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by lasote** [comment]
Hi! It is a challenging feature but we are doing steps to make the local cache concurrent and being able to do things like that. But we are not there yet. We are introducing a new cache model in Conan 2.0 so we might be able to see progress then.

In the meantime, just in case you don't know there is an environment variable [CONAN_CPU_COUNT](https://docs.conan.io/en/latest/reference/env_vars.html?highlight=cores#conan-cpu-count) that might help to use more cores in a single build.

**Comment by blackliner** [comment]
hey @lasote , thanks for the heads up. I know about that variable, but as we try to stick to the `conan` philosphy of one library per package, those tend to be very small packages with only a few `*.cpp` files, which do not really make my 16 core workstation sweat 😅 

So the idea would be to use the power to build those small packages in parallel, hiding all the output (which is shown in case a build fails of course) and showing a nice info for the overall progress.

**Comment by blackliner** [comment]
Ok, working on a wrapper script that uses a custom dependency tree to build multiple packages in parallel. Hitting the concurrency limitations conan has in the local cache.

If two packages get build in parallel (separate conan create processes), and they depend on the same package that can be downloaded, a race condition can happen. If the first process is running the build while the second process is cleaning up the destination path for the dependency, the first process fails due to missing header files etc.

So even if there is limited concurrency support for packages that get build, it seems that the download process is not covered by the read/write mutex. Any advice? We have over 350 packages, and if a common package is changed at the top of the dependency tree, we have to rebuild (+ test + lint + sonarqube + helix qac + ...) 350 packages sequentially :-(

**Comment by blackliner** [comment]
One idea I had was to try and create a separate conan cache for each build, like building in a sandbox. But I assume we lose most of the parallelization gains when dependencies have to be recreated all the time.

What kind of works for now is to have 5 retries in case a build fails, in the hope that the race condition does not occur 🙈

**Comment by blackliner** [comment]
Hm, maybe https://docs.conan.io/en/latest/uploading_packages/running_your_server.html can help out as a quick way to up/download between those builds 🤔 Does the server support parallel read write access?

**Comment by memsharded** [comment]
Maybe trying to use a different cache per parallel build, but using the file-download cache https://docs.conan.io/en/latest/configuration/download_cache.html to avoid hitting too much the server could be an improvement in the meantime?

**Comment by blackliner** [comment]
@memsharded we are already using the file-download cache. My assumption is that its not the download cache that is the issue (that has proven to have a good concurrency capability in our CI, as we share the same download cache on the host node with multiple pods running concurrently), but the followup install process that first cleans the dir in the local cache, then unpacks the prebuilt binary and headers.

**Comment by memsharded** [comment]
Closing in favor of https://github.com/conan-io/conan/issues/15840 where the Conan 2 cache concurrency future progress will be reported.
